optuna/_optimize.py

```diff
@@ -0,0 +1,240 @@
+                    warnings.warn(
+                        "The default storage cannot be shared by multiple processes. "
+                        "Please use an RDB (RDBStorage) when you use joblib for "
+                        "multi-processing. The usage of RDBStorage can be found in "
+                        "https://optuna.readthedocs.io/en/stable/tutorial/rdb.html.",
+                        UserWarning,
+                    )
+
+                parallel(
+                    delayed(_optimize_sequential)(
+                        study,
+                        func,
+                        1,
+                        timeout,
+                        catch,
+                        callbacks,
+                        gc_after_trial,
+                        reseed_sampler_rng=True,
+                        time_start=time_start,
+                        progress_bar=None,
+                    )
+                    for _ in _iter
+                )
+    finally:
+        study._optimize_lock.release()
+        progress_bar.close()
+
+
+def _optimize_sequential(
+    study: "optuna.Study",
+    func: "optuna.study.ObjectiveFuncType",
+    n_trials: Optional[int],
+    timeout: Optional[float],
+    catch: Tuple[Type[Exception], ...],
+    callbacks: Optional[List[Callable[["optuna.Study", FrozenTrial], None]]],
+    gc_after_trial: bool,
+    reseed_sampler_rng: bool,
+    time_start: Optional[datetime.datetime],
+    progress_bar: Optional[pbar_module._ProgressBar],
+) -> None:
+    if reseed_sampler_rng:
+        study.sampler.reseed_rng()
+
+    i_trial = 0
+
+    if time_start is None:
+        time_start = datetime.datetime.now()
+
+    while True:
+        if study._stop_flag:
+            break
+
+        if n_trials is not None:
+            if i_trial >= n_trials:
+                break
+            i_trial += 1
+
+        if timeout is not None:
+            elapsed_seconds = (datetime.datetime.now() - time_start).total_seconds()
+            if elapsed_seconds >= timeout:
+                break
+
+        try:
+            trial = _run_trial(study, func, catch)
+        except Exception:
+            raise
+        finally:
+            # The following line mitigates memory problems that can be occurred in some
+            # environments (e.g., services that use computing containers such as CircleCI).
+            # Please refer to the following PR for further details:
+            # https://github.com/optuna/optuna/pull/325.
+            if gc_after_trial:
+                gc.collect()
+
+        if callbacks is not None:
+            frozen_trial = copy.deepcopy(study._storage.get_trial(trial._trial_id))
+            for callback in callbacks:
+                callback(study, frozen_trial)
+
+        if progress_bar is not None:
+            progress_bar.update((datetime.datetime.now() - time_start).total_seconds())
+
+    study._storage.remove_session()
```

Comment:
I believe we should have this method called when exceptions are raised and propagated from `_run_trial` as well. It's sort of orthogonal to this PR.